design/observerability.md

```diff
@@ -1,150 +1,263 @@
+- Defining metrics required by the proposed dashboard as well as designing a mechanism for exporting them to prometheus.
+- Design logging improvements for better error tracking.
 
-## Types of Observerability
+What's out of the scope:
+- Traceability - tracking a user across the whole authorization flow. The progress within this space is tracked in
+[KEYCLOAK-7432](https://issues.redhat.com/browse/KEYCLOAK-7432).
 
-Health - a health check endpoint returns a simple binary up or down state of the system. This is very useful in managed
-environments like Kubernetes and OpenShift where the controller can automatically stop broken instances.
+## Coarse-grained Keycloak Services
 
-Metrics - metrics provide a more detailed view of the health and state of the system. From metrics it is possible to
-find out how the system is behaving, not just if the system is up or down. Examples here could be observing a large
-amount of error responses or response times increasing beyond a given threshold.
+For the purpose of this design, Keycloak has been divided into the following Services:
+* Admin - both the Console and the REST endpoints.
+* Account - both the Console and the REST endpoints.
+* OIDC - All endpoints for OpenID Connect.
+* Token Exchange - Token Exchange endpoint.
+* SAML - All endpoints for SAML spec.
+* User Federation - LDAP servers used for User Federation
 
-Logging - metrics provides the details on how the system is behaving, but will not always be able to provide sufficient
-information to debug the system. This is where logging comes in. Through well defined logging it is possible to further
-obtain information on how the system is behaving and resolving potential issues.
+[^1]: The Token Exchange feature belongs to oAuth spec, so it can't go under OIDC.
 
-Tracing - in a distributed system it can be difficult to find the journey a single transactions takes through the
-systems. With distributed tracing it is possible to do exactly this.
+## SRE Dashboard
 
+![SRE Dashboard](img/observability-dashboard.png)
 
-## Observerability in WildFly
+NOTE: The diagram implementation and specific charts might change during the implementation phase. The above proposal
+shows only the general idea.
 
-WildFly 16 comes with built-in support for Observerability, which will make it relatively straightforward to add the same
-to Keycloak.
+The proposal consists of 3 sections:
+- System health - Enables the SRE Team to quickly tell if Keycloak cluster is up and running. This section consists of:
+  - Readiness and Liveness Probes chart
+  - HTTP 5xx error rate for all major Keycloak Services, including: Admin Console, Account Console, OIDC, SAML
+  - Aggregated latency across the most important endpoints
+  - (Optional) Error rate in logs [^2]
+- Detailed Metrics - This section helps foresee upcoming failures. It includes:
+  - Pod memory, CPU and container storage charts
+  - GC Times
+  - Heap and non-heap memory metrics
```

Comment:
Per-cache utilization should be here as a separate item for the time being to display whether configuration needs tweaking.

Reply:
Please note, that the new Observability initiative targets Quarkus deployment, which by definition will use a remote Infinispan cluster. Having this in mind, I'm not sure if monitoring Infinispan caches should be our concern. Perhaps this is something for the Infinispan Observablity? @ryanemerson - do you have any thoughts around this?